File: backend/app/llm/runtime.py
"""LLM 运行时配置解析。"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Literal, cast

LLMMode = Literal["deterministic", "provider"]

_LLM_MODE_ENV_KEY = "LLM_MODE"
_LLM_MODEL_ENV_KEY = "LLM_MODEL"
_LLM_API_KEY_ENV_KEY = "LLM_API_KEY"
_LLM_BASE_URL_ENV_KEY = "LLM_BASE_URL"


@dataclass(slots=True, frozen=True)
class LLMConfig:
    """LLM 配置快照。"""

    mode: LLMMode
    model: str
    api_key: str | None
    base_url: str | None


def resolve_llm_config(env: dict[str, str] | None = None) -> LLMConfig:
    """解析 LLM 配置，默认 deterministic 以保证测试可复现。"""
    env_map = env or os.environ
    mode_raw = env_map.get(_LLM_MODE_ENV_KEY, "deterministic").strip().lower()
    if mode_raw not in {"deterministic", "provider"}:
        raise ValueError("LLM_MODE must be one of: deterministic, provider")

    model = env_map.get(_LLM_MODEL_ENV_KEY, "gpt-4o-mini").strip() or "gpt-4o-mini"
    api_key = env_map.get(_LLM_API_KEY_ENV_KEY, "").strip() or None
    base_url = env_map.get(_LLM_BASE_URL_ENV_KEY, "").strip() or None

    if mode_raw == "provider" and api_key is None:
        raise ValueError("LLM_API_KEY is required when LLM_MODE=provider")

    return LLMConfig(
        mode=cast(LLMMode, mode_raw),
        model=model,
        api_key=api_key,
        base_url=base_url,
    )

File: backend-hono/src/llm/runtime.ts
/**
 * LLM 运行时配置解析。
 * WHAT: 统一解析 deterministic/provider 模式与 OpenAI 兼容端点参数。
 * WHY: 让路由层不关心环境变量细节，避免同一配置逻辑在多处分叉。
 */
export type LLMMode = 'deterministic' | 'provider';

export type LLMRuntimeEnv = {
  LLM_MODE?: string;
  LLM_MODEL?: string;
  LLM_API_KEY?: string;
  LLM_BASE_URL?: string;
  OPENAI_API_KEY?: string;
  OPENAI_BASE_URL?: string;
};

export type LLMConfig = {
  mode: LLMMode;
  model: string;
  apiKey: string | null;
  baseURL: string | null;
};

function normalizeOptional(value?: string) {
  if (!value) {
    return null;
  }
  const normalized = value.trim();
  return normalized.length > 0 ? normalized : null;
}

export function resolveLLMConfig(env: LLMRuntimeEnv): LLMConfig {
  const modeRaw = (env.LLM_MODE ?? 'deterministic').trim().toLowerCase();
  if (modeRaw !== 'deterministic' && modeRaw !== 'provider') {
    throw new Error('LLM_MODE must be one of: deterministic, provider');
  }

  const model = normalizeOptional(env.LLM_MODEL) ?? 'gpt-4o-mini';
  const apiKey = normalizeOptional(env.LLM_API_KEY) ?? normalizeOptional(env.OPENAI_API_KEY);
  const baseURL = normalizeOptional(env.LLM_BASE_URL) ?? normalizeOptional(env.OPENAI_BASE_URL);

  if (modeRaw === 'provider' && apiKey === null) {
    throw new Error('LLM_API_KEY is required when LLM_MODE=provider');
  }

  return {
    mode: modeRaw,
    model,
    apiKey,
    baseURL
  };
}
